src/Common/Scheduler/CPULeaseAllocation.cpp

```diff
@@ -0,0 +1,533 @@
+    };
+
+    if (timeout == std::chrono::milliseconds::max())
+    {
+        threads.wake[thread_num].wait(lock, predicate);
+        return true; // Granted
+    }
+    else
+    {
+        return threads.wake[thread_num].wait_for(lock, timeout, predicate);
+    }
+}
+
+void CPULeaseAllocation::consume(std::unique_lock<std::mutex> & lock, ResourceCost delta_ns)
+{
+    consumed_ns += delta_ns;
+    if (allocated > 0 && consumed_ns >= tail->max_consumed)
+    {
+        --allocated;
+        --granted;
+        if (granted <= 0 && !exception)
+            acquirable.store(false, std::memory_order_relaxed);
+        tail->finish();
+        if (tail->is_master_slot)
+            request_master_slot = true;
+        ++tail;
+        if (tail == requests.end())
+            tail = requests.begin();
+        LOG_EVENT(C);
+        if (!enqueued)
+            schedule(lock); // In case if we renew the last slot, otherwise the next request is already scheduled
+        // NOTE: we do not finish more than one request per one report to avoid stalling the pipeline for reports larger than quantum
+    }
+}
+
+void CPULeaseAllocation::schedule(std::unique_lock<std::mutex> & lock)
+{
+    chassert(!enqueued);
+
+    if (allocated == max_threads || shutdown)
+        return;
+
+    ResourceCost cost = settings.quantum_ns + std::max<ResourceCost>(0, consumed_ns - requested_ns);
+    requested_ns += cost;
+    head->reset(cost);
+    head->is_master_slot = std::exchange(request_master_slot, false);
+    head->max_consumed = requested_ns; // Lease expires if we consume what we requested
+
+    if (auto * queue = head->is_master_slot ? master_link.queue : worker_link.queue)
+    {
+        head->is_noncompeting = false;
+        // We do not use enqueueRequestUsingBudget() because it redistributes resource between requests in the queue (which might be from different queries).
+        // Instead we do budgeting for every query independently for better fairness
+        queue->enqueueRequest(&*head);
+        enqueued = true;
+        scheduled_increment.add();
+        wait_timer.emplace(CurrentThread::getProfileEvents().timer(ProfileEvents::ConcurrencyControlWaitMicroseconds));
+        LOG_EVENT(E);
+    }
+    else // noncompeting slot - provide immediately for free
+    {
+        head->is_noncompeting = true;
+        grantImpl(lock);
+    }
+}
+
+void CPULeaseAllocation::release(Lease & lease)
+{
+    UInt64 thread_time_ns = clock_gettime_ns(CLOCK_THREAD_CPUTIME_ID);
+    chassert(thread_time_ns >= lease.last_report_ns); // This is guaranteed on Linux for thread clock
+    ResourceCost delta_ns = thread_time_ns - lease.last_report_ns;
+    lease.last_report_ns = thread_time_ns;
+
+    // Report the last chunk of consumed resource
+    std::unique_lock lock{mutex};
+    consume(lock, delta_ns);
+
+    // Release the slot
+    downscale(lease.slot_id);
+}
+
+bool CPULeaseAllocation::isRequesting() const
+{
+    std::unique_lock lock{mutex};
```

Comment:
I think `lock_guard` is better for similar cases

Reply:
Fixed